### ISSUE TITLE
Analyzer for the hard-to-spot expr?.Prop.ShouldBe(x)

### ISSUE THREAD

**Issue by jnm2** [shouldly/shouldly#710]
`?.` is a natural thing to reach for here, but it doesn't have the intended effect when `expr` is null:

```cs
expr?.Prop1.ShouldBe(x);
expr?.Prop2.ShouldBe(y);
```

I can't think of a way to address this without starting to ship an analyzer with the Shouldly package.

This was probably the intended effect:

```cs
(expr?.Prop1).ShouldBe(x);
(expr?.Prop2).ShouldBe(y);
```

A fix that generates assertion messages with better information content would be to use our new chaining ability (requires Shouldly v4+):

```cs
expr.ShouldNotBeNull().Prop1.ShouldBe(x);
expr.ShouldNotBeNull().Prop2.ShouldBe(y);
```

This is not a good option. 1. The `!` is required if you use NRTs or you don't use NRTs but you do have ReSharper, and you want to avoid squiggles. 2. The test fails with InvalidOperationException and no expression written out instead of getting a nice Shouldly error.

```cs
expr!.Value.Prop1.ShouldBe(x);
expr!.Value.Prop2.ShouldBe(y);
```

What are your thoughts on maintaining an analyzer?

**Comment by SimonCropp** [comment]
i have not maintained an analyzer. but i dont object to one being added

**Comment by JosephWoodward** [comment]
> What are your thoughts on maintaining an analyzer?

It's something I've considered before, and even started trying to make one. I'd be a great way to start to add some additional functionality into Shouldly (I was thinking, being able to switch simple xunit tests to Shouldly assertions for instance).

**Comment by JosephWoodward** [comment]
@jnm2 Have you built an Analyzer before?

**Comment by jnm2** [comment]
Yes, a few from scratch and contributing to many others. I'd be happy to set everything up.

One question to answer would be whether to introduce a new package, Shouldly.Analyzers, or whether to put the analyzers in the same package. Most libraries go the separate package route because it allows independent control of the versions which can be useful. It's also easier to use the SDK to create the package when they are separate, but this is a hurdle we can easily overcome.

We can still make the Shouldly package depend on the Shouldly.Analyers package if we want analyzers to be opt-out instead of opt-in, which is something I prefer. Otherwise, as a consumer, it usually won't occur to me to check for and remember to always install the analyzer package along with the library package.

What are your thoughts?
